bundles/admin/admin-layereditor/view/AdminLayerForm/VisualizationTabPane/VectorStyle.jsx

```diff
@@ -16,78 +17,170 @@ const FullWidthSpace = styled(Space)`
+        return 'cesium';
+    }
+    if (type === 'vectortilelayer') {
+        return 'mapbox';
+    }
+    throw new Error(`Unknown layer type (${type}). Tried to get external vector style type)`);
+};
+const parseJsonStyle = (mode, style) => {
+    let parsed = {};
+    const parse = json => {
+        try {
+            return JSON.parse(json);
+        } catch { /* ignored */ }
+    };
+    if (mode === 'json') {
+        const featureStyle = parse(style.featureStyle);
+        if (typeof featureStyle === 'object') {
+            parsed.featureStyle = featureStyle;
+        }
+        const optionalStyles = parse(style.optionalStyles);
+        if (Array.isArray(optionalStyles)) {
+            parsed.optionalStyles = optionalStyles;
+        }
+    } else if (mode === 'external') {
+        const ext = parse(style);
+        if (typeof ext === 'object') {
+            parsed = ext;
+        }
+    }
+    return Object.keys(parsed).length ? parsed : null;
+};
+const stringify = (json) => {
+    try {
+        return JSON.stringify(json) || '';
+    } catch { /* ignored */ }
+    return '';
 };
+const getRuntimeId = () => Date.now().valueOf(); // Long in backend
 
-export const VectorStyle = LocaleConsumer((props) => {
-    const newStyleName = props.getMessage('styles.vector.newStyleName');
-    const [editorState, setEditorState] = useState({
-        modalVisibility: false,
-        currentStyle: {},
-        styleId: '',
-        styleName: newStyleName,
-        originalName: '',
-        validates: false
+export const VectorStyle = LocaleConsumer(({ layer, getMessage, controller, external }) => {
+    const newStyleName = getMessage('styles.vector.newStyleName');
+    const [state, setState] = useState({
+        modal: null
     });
+    const hasValidName = () => typeof state.name === 'string' && state.name.trim().length > 0;
 
-    const saveStyle = () => props.controller.saveStyleToLayer(editorState.currentStyle, editorState.styleName, editorState.styleId);
-    const onModalCancel = () => setEditorState({ ...editorState, modalVisibility: false });
-    const resetNewStyle = () => setEditorState({ ...editorState, styleId: '', styleName: newStyleName, originalName: '', currentStyle: {}, modalVisibility: true });
-    const setName = (name) => setEditorState({ ...editorState, styleName: name });
+    const onModalClose = () => setState({ modal: null });
+    const setName = (name) => setState({ ...state, name });
+    const updateStyle = (style) => setState({ ...state, style });
+    const addStyle = (mode, template) => setState({ id: getRuntimeId(), name: newStyleName, style: template, modal: mode });
+    const editStyle = (mode, vectorStyle) => {
+        const { style, ...state } = vectorStyle;
+        if (mode === 'json') {
+            if (state.type === 'oskari') {
+                const { featureStyle, optionalStyles } = style;
+                state.style = {
+                    featureStyle: stringify(featureStyle),
```

Comment:
Is there some reason why these are stringified?

Reply:
If I remember right, it is stringified for json text area input. And maybe to beautify json but it looks like space param is missing from actual JSON.stringify.